### PR TITLE
New resource: `azurerm_kusto_table`

### LINKED ISSUE #6392: Support for creating Azure Data Explorer tables and ingestion mappings.
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Support for creating Azure Data Explorer tables and ingestion mappings. This is required to enable e2e automated deployment of Azure Data Explorer solution that includes creating cluster, creating database, setting database permissions, creating tables, creating ingestion mappings, and creating data connections.  

### New or Affected Resource(s)

* azurerm_kusto

### References

Related Issue: https://github.com/terraform-providers/terraform-provider-azurerm/issues/5553
Documentation: [Create table](https://docs.microsoft.com/azure/kusto/management/tables), [Create data mappings](https://docs.microsoft.com/azure/kusto/management/mappings)

### PR DESCRIPTION
**Pull request by jrauschenbusch**
Fixes #6392

Coded in collaboration with @njuCZ (Thx for your commitment for the auth related code)

```sh
$ make acctests SERVICE='kusto' TESTARGS='-run=TestAccAzureRMKustoTable' TESTTIMEOUT='60m' 
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/kusto/tests/ -run=TestAccAzureRMKustoTable -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMKustoTable_basic
=== PAUSE TestAccAzureRMKustoTable_basic
=== RUN   TestAccAzureRMKustoTable_update
=== PAUSE TestAccAzureRMKustoTable_update
=== RUN   TestAccAzureRMKustoTable_requiresImport
=== PAUSE TestAccAzureRMKustoTable_requiresImport
=== RUN   TestAccAzureRMKustoTable_complete
=== PAUSE TestAccAzureRMKustoTable_complete
=== CONT  TestAccAzureRMKustoTable_basic
=== CONT  TestAccAzureRMKustoTable_complete
=== CONT  TestAccAzureRMKustoTable_requiresImport
=== CONT  TestAccAzureRMKustoTable_update
--- PASS: TestAccAzureRMKustoTable_basic (1413.65s)
--- PASS: TestAccAzureRMKustoTable_complete (1526.54s)
--- PASS: TestAccAzureRMKustoTable_requiresImport (1539.01s)
--- PASS: TestAccAzureRMKustoTable_update (1709.70s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/kusto/tests 1711.134s
```

### CODE REVIEW

File: azurerm/internal/services/kusto/kusto_table_resource.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+func resourceArmKustoTable() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmKustoTableCreateUpdate,
+		Read:   resourceArmKustoTableRead,
+		Update: resourceArmKustoTableCreateUpdate,
+		Delete: resourceArmKustoTableDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
```
Comment (njuCZ) [inline azurerm/internal/services/kusto/kusto_table_resource.go:31]: we should check the id format just like other resources do

File: azurerm/internal/services/kusto/tests/kusto_table_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+			{
+				Config: testAccAzureRMKustoTable_basic(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMKustoTableExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+		},
+	})
+}
```
Comment (njuCZ) [inline azurerm/internal/services/kusto/tests/kusto_table_resource_test.go:35]: in general, we will have basic, complete and update test, could you please add more tests

File: website/docs/r/kusto_table.html.markdown [hunk 3]
```diff
@@ -0,0 +1,72 @@
+---
+subcategory: "Data Explorer"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_kusto_table"
+description: |-
+  Manages a Data Explorer.
```
Comment (njuCZ) [inline website/docs/r/kusto_table.html.markdown:6]:
```suggestion
  Manages a Data Explorer Table.
```

File: website/docs/r/kusto_table.html.markdown [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+---
+
+# azurerm_kusto_table
+
+Manages a Data Explorer.
```
Comment (njuCZ) [inline website/docs/r/kusto_table.html.markdown:11]: above

File: website/docs/r/kusto_table.html.markdown [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+```hcl
+resource "azurerm_kusto_table" "example" {
+  name = "example"
+  database_id = "TODO"
+
+  column {
+    name = "example"
+    type = "TODO"    
+  }
+}
```
Comment (njuCZ) [inline website/docs/r/kusto_table.html.markdown:24]: in general, we will list all the dependency in the example usage and remember to format it. you could use the tool: terrafmt

File: website/docs/r/kusto_table.html.markdown [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `column` - (Required) One or more `column` blocks as defined below.
+
+* `database_id` - (Required) The ID of the TODO. Changing this forces a new Data Explorer to be created.
+
+* `name` - (Required) The name which should be used for this Data Explorer. Changing this forces a new Data Explorer to be created.
```
Comment (njuCZ) [inline website/docs/r/kusto_table.html.markdown:35]:
```suggestion
* `name` - (Required) The name which should be used for this Data Explorer Table. Changing this forces a new Data Explorer to be created.
```

File: website/docs/r/kusto_table.html.markdown [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+* `id` - The ID of the Data Explorer.
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/docs/configuration/resources.html#timeouts) for certain actions:
+
+* `create` - (Defaults to 1 hour) Used when creating the Data Explorer.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Data Explorer.
+* `update` - (Defaults to 1 hour) Used when updating the Data Explorer.
+* `delete` - (Defaults to 1 hour) Used when deleting the Data Explorer.
```
Comment (njuCZ) [inline website/docs/r/kusto_table.html.markdown:64]:
```suggestion
* `create` - (Defaults to 1 hour) Used when creating the Data Explorer Table.
* `read` - (Defaults to 5 minutes) Used when retrieving the Data Explorer Table.
* `update` - (Defaults to 1 hour) Used when updating the Data Explorer Table.
* `delete` - (Defaults to 1 hour) Used when deleting the Data Explorer Table.
```

Comment (jrauschenbusch) [reply website/docs/r/kusto_table.html.markdown]: done

File: website/docs/r/kusto_table.html.markdown (hunk 6 above)
Comment (jrauschenbusch) [reply website/docs/r/kusto_table.html.markdown]: done

File: website/docs/r/kusto_table.html.markdown (hunk 5 above)
Comment (jrauschenbusch) [reply website/docs/r/kusto_table.html.markdown]: fixed

File: website/docs/r/kusto_table.html.markdown (hunk 4 above)
Comment (jrauschenbusch) [reply website/docs/r/kusto_table.html.markdown]: fixed

File: website/docs/r/kusto_table.html.markdown (hunk 3 above)
Comment (jrauschenbusch) [reply website/docs/r/kusto_table.html.markdown]: fixed

File: azurerm/internal/services/kusto/kusto_table_resource.go (hunk 1 above)
Comment (jrauschenbusch) [reply azurerm/internal/services/kusto/kusto_table_resource.go:31]: Could you be more precise what you would change here?

Comment (njuCZ) [reply azurerm/internal/services/kusto/kusto_table_resource.go:31]: for example: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/appplatform/spring_cloud_app_resource.go#L28-L31

Comment (jrauschenbusch) [reply azurerm/internal/services/kusto/kusto_table_resource.go]: Fixed

**Comment by jrauschenbusch** [comment]
@njuCZ I am currently thinking about whether the `.alter-table` command would be better for the update strategy of the table resource. This would mean that there would be a risk of columns being accidentally deleted (if the terraform plan is not taken into account), but from my point of view it would represent the table better as a resource, since the created / changed schema would really correspond to what is defined via the HCL markup. What is your opinion?

see following links:
- https://docs.microsoft.com/de-de/azure/data-explorer/kusto/management/create-merge-table-command
- https://docs.microsoft.com/de-de/azure/data-explorer/kusto/management/alter-table-command

File: azurerm/internal/services/kusto/tests/kusto_table_resource_test.go (hunk 2 above)
Comment (jrauschenbusch) [reply azurerm/internal/services/kusto/tests/kusto_table_resource_test.go]: done

**Comment by jrauschenbusch** [comment]
> @njuCZ I am currently thinking about whether the `.alter table` command would be better for the update strategy of the table resource. This would mean that there would be a risk of columns being accidentally deleted (if the terraform plan is not taken into account), but from my point of view it would represent the table better as a resource, since the created / changed schema would really correspond to what is defined via the HCL markup. What is your opinion?
> 
> see following links:
> 
> * https://docs.microsoft.com/de-de/azure/data-explorer/kusto/management/create-merge-table-command
> * https://docs.microsoft.com/de-de/azure/data-explorer/kusto/management/alter-table-command

I had to switch to an dedicated update strategy, as otherwise tests were failing or terraform's state handling was not working properly. I already updated all resources and tests. It seems that now everything is working fine.

**Comment by jrauschenbusch** [comment]
@mbfrahry Request for review

**Comment by njuCZ** [comment]
@jrauschenbusch there is conflicts for this PR. We should fix all of them before requesting hashicorp guys to review

**Comment by jrauschenbusch** [comment]
> @jrauschenbusch there is conflicts for this PR. We should fix all of them before requesting hashicorp guys to review

Now the PR should be in a presentable state.

**Comment by jrauschenbusch** [comment]
Is there any chance that this PR will get reviewed?

**Comment by alxy** [comment]
@jrauschenbusch Maybe it makes sense to extract that to a separate provider? I can sort of understand that the maintainers of this repo don't want to go too far down the road with resources they support, as this really cannot be considered an Azure resource. However, ADX/kusto is clearly lacking a good tool to handle deployments of the data schemas (tables, functions), so it might be a good candidate for a custom provider to support that. What do you think?

**Comment by jrauschenbusch** [comment]
@alxy Sorry for the late response. I think you're right. It should be a separate terraform provider. Unfortunately there was no single feedback from the core Terraform AzureRM team. I have to find a free time slot where i can start with the implementation. Stay tuned.

**Comment by katbyte** [comment]
@jrauschenbusch - id the plan is to create a new provider for kusto can this can be closed in favour of it then?

**Comment by jrauschenbusch** [comment]
@katbyte I'll close it now. I just wanted to notice everyone who is interested on it where the provider will be located. But did not found any time until now due to the birth of my son. But i did not forget it. It's postponed but not cancelled.

**Pull request closed without merging**